File: Roadmap/21 - CALLBACKS/python/cesar-ch.py
"""
    * 20 PETICIONES HTTP 
"""


def greet(name, callback):
    print(f"Hello, {name}!")
    callback()


def say_goodbye():
    print("Goodbye!")


greet("Mouredev", say_goodbye)


"""
    * DIFICULTAD EXTRA
"""
import time
import random


def procesar_pedido(plato, confirmacion_callback, listo_callback, entrega_callback):
    print(f"Procesando pedido para: {plato}")
    confirmacion_callback(plato)

    tiempo_preparacion = random.randint(1, 10)
    print("El tiempo de preparacion es: " + str(tiempo_preparacion))
    time.sleep(tiempo_preparacion)
    listo_callback(plato)

    tiempo_entrega = random.randint(1, 10)
    print("El tiempo de entrega es: " + str(tiempo_entrega))
    time.sleep(tiempo_entrega)
    entrega_callback(plato)


def confirmar_pedido(plato):
    print(f"Pedido confirmado para: {plato}")


def plato_listo(plato):
    print(f"El plato {plato} está listo para ser entregado.")


def entregar_pedido(plato):
    print(f"El plato {plato} ha sido entregado.")


if __name__ == "__main__":
    procesar_pedido("Pizza", confirmar_pedido, plato_listo, entregar_pedido)

File: Roadmap/21 - CALLBACKS/javascript/cesar-ch.js
/*
    * #21 CALLBACKS
*/

function greet(name, callback) {
    console.log("Hello, " + name + "!");
    callback();
}

function sayGoodbye() {
    console.log("Goodbye!");
}

greet("Mouredev", sayGoodbye);

/*
    * DIFICULTAD EXTRA
*/

function procesarPedido(plato, confirmacionCallback, listoCallback, entregaCallback) {
    console.log(`Procesando pedido para: ${plato}`);
    confirmacionCallback(plato);

    const tiempoPreparacion = Math.floor(Math.random() * 10) + 1;
    console.log("Tiempo de preparacion es: " + tiempoPreparacion)
    setTimeout(() => {
        listoCallback(plato);

        const tiempoEntrega = Math.floor(Math.random() * 10) + 1;
        console.log("Tiempo de entrega es: " + tiempoEntrega)
        setTimeout(() => {
            entregaCallback(plato);
        }, tiempoEntrega);
    }, tiempoPreparacion);
}

function confirmarPedido(plato) {
    console.log(`Pedido confirmado para: ${plato}`);
}

function platoListo(plato) {
    console.log(`El plato ${plato} está listo para ser entregado.`);
}

function entregarPedido(plato) {
    console.log(`El plato ${plato} ha sido entregado.`);
}

procesarPedido("Pizza", confirmarPedido, platoListo, entregarPedido);
